packages/shared/babel-ast-utils/src/index.js

```diff
@@ -24,6 +25,42 @@ import {generator, expressionsPrecedence} from './generator';
 
 export {babelErrorEnhancer};
 
+export function remapAstLocations(ast: BabelNodeFile, map: SourceMap) {
+  // remap ast to original mappings
+  // This improves sourcemap accuracy and fixes sourcemaps when scope-hoisting
+  traverseAll(ast.program, node => {
+    if (node.loc) {
+      if (node.loc?.start) {
+        let mapping = map.findClosestMapping(
+          node.loc.start.line,
+          node.loc.start.column,
+        );
+
+        if (mapping?.original) {
+          // $FlowFixMe
+          node.loc.start.line = mapping.original.line;
+          // $FlowFixMe
+          node.loc.start.column = mapping.original.column;
+
+          // $FlowFixMe
+          let length = node.loc.end.column - node.loc.start.column;
```

Comment:
This is a bit hacky, but the alternative I've thought of isn't much better and probably slower which is using the mapping after the found mapping to get the end position. This would also not be a lot more accurate as sourcemaps are inherently not very accurate to begin with.